astropy/cosmology/core.py

```diff
@@ -166,23 +460,101 @@ def __eq__(self, other):
+
+        Two cosmologies may be equivalent even if not the same class.
+        For example, an instance of ``LambdaCDM`` might have :math:`\Omega_0=1`
+        and :math:`\Omega_k=0` and therefore flat, like ``FlatLambdaCDM``.
+
+        Parameters
+        ----------
+        other : `~astropy.cosmology.Cosmology` subclass instance
+            The object in which to compare.
 
+        Returns
+        -------
+        bool
+            True if cosmologies are equivalent, False otherwise.
         """
+        # The options are: 1) same class & parameters; 2) same class, different
+        # parameters; 3) different classes, equivalent parameters; 4) different
+        # classes, different parameters. (1) & (3) => True, (2) & (4) => False.
+
         if not isinstance(other, Cosmology):
             return False
+        # passed this basic check, so 'other' should have method '__equiv__'
 
-        sias, oias = self._init_arguments, other._init_arguments
+        equiv = self.__equiv__(other)  # first try from 'self'
 
-        # check if the cosmologies have identical signatures.
-        # this protects against one cosmology having a superset of input
-        # parameters to another cosmology.
-        if (sias.keys() ^ oias.keys()) - {'meta'}:
-            return False
+        if equiv is NotImplemented:  # that failed, try starting from 'other'
+            equiv = other.__equiv__(self)
+
+        return equiv if equiv is not NotImplemented else False
+
+    @abc.abstractmethod
+    def __equiv__(self, other):
+        """
+        Cosmology equivalence.
+        Use method ``is_equivalent`` for actual check!
+        Must be overridden in subclasses.
+
+        Returns
+        -------
+        `NotImplemented`
+
+        """
+        return NotImplemented
+
+    # -----------------------------------------------------
+
+    def __repr__(self):
+        ps = self.info("parameters")  # values
+        cps = self.__class__.info("parameters")  # Parameter objects
+
+        namelead = f"{self.__class__.__qualname__}("
+        if self.name is not None:
+            namelead += f"name=\"{self.name}\", "
+        # nicely formatted parameters
+        fmtps = (k + '=' + format(v, cps[k].format_spec if v is not None else '')
+                 for k, v in ps.items())
+
+        return namelead + ", ".join(fmtps) + ")"
 
-        # are all the non-excluded immutable arguments equal?
-        return all((np.all(oias[k] == v) for k, v in sias.items()
-                    if k != "meta"))
+
+class FlatCosmologyMixin(metaclass=abc.ABCMeta):
+    """
+    Mixin class for flat cosmologies. Do NOT instantiate directly.
+    Note that all instances of ``FlatCosmologyMixin`` are flat, but not all
+    flat cosmologies are instances of ``FlatCosmologyMixin``. As example,
+    ``LambdaCDM`` **may** be flat (for the a specific set of parameter values),
+    but ``FlatLambdaCDM`` **will** be flat.
+    """
+
+    @abc.abstractmethod
```

Comment:
It still seems this should not be necessary!

Reply:
With the suggestion for ``__equiv__ = __eq__``, it is not 👍